### PR TITLE
Add the ability to pause & unpause DMs

### LINKED ISSUE #182: Todo: Pause DMs command

### PR DESCRIPTION
**Pull request by FloatingMilkshake**
This PR closes #182.

Adds the ability to pause and unpause DMs. This is the same as pausing DMs from a server's "Security Actions" menu:
image

Because of D#+ limitations, the bot sends manual HTTP requests to Discord's API. It feels a bit hacky, but it works!

'Classic' message commands and slash commands are both added, to both pause DMs:
image
...and unpause them:
image
image

The commands fail with a warning if an invalid time is provided when attempting to pause DMs, or if DMs are already unpaused when attempting to unpause them.
image

The slash commands are limited to moderators only (i.e. not trial moderators). Because of limitations, they are restricted with the 'Moderate Members' permission, so they will be visible to trial moderators, but unusable.
image

### CODE REVIEW

File: Commands/InteractionCommands/SecurityActionInteractions.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+            // set actions
+            var setActionsResponse = await SecurityActionHelpers.SetCurrentSecurityActions(ctx.Guild.Id, newSecurityActions);
+
+            // respond
+            if (setActionsResponse.IsSuccessStatusCode)
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully paused DMs for **{TimeHelpers.TimeToPrettyFormat(t.Subtract(ctx.Interaction.CreationTimestamp.DateTime.ToLocalTime()), false)}**!");
+            else
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to pause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
+        }
```
Comment (Erisa) [inline Commands/InteractionCommands/SecurityActionInteractions.cs:48]: Lets log it
```suggestion
            else
            {
                ctx.Client.Logger.LogError("Failed to set Security Actions.\nPayload: {payload}\nResponse: {statuscode} {body}", newSecurityActions.ToString(), (int)setActionsResponse.StatusCode, await setActionsResponse.Content.ReadAsStringAsync());
                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to unpause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
            }
```

File: Commands/InteractionCommands/SecurityActionInteractions.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            
+            // set actions
+            var setActionsResponse = await SecurityActionHelpers.SetCurrentSecurityActions(ctx.Guild.Id, newSecurityActions);
+
+            // respond
+            if (setActionsResponse.IsSuccessStatusCode)
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully unpaused DMs!");
+            else
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to unpause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
+        }
```
Comment (Erisa) [inline Commands/InteractionCommands/SecurityActionInteractions.cs:92]: Lets log it
```suggestion
            else
            {
                ctx.Client.Logger.LogError("Failed to set Security Actions.\nPayload: {payload}\nResponse: {statuscode} {body}", newSecurityActions.ToString(), (int)setActionsResponse.StatusCode, await setActionsResponse.Content.ReadAsStringAsync());
                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to pause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
            }
```

File: Commands/InteractionCommands/SecurityActionInteractions.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                invites_disabled_until = invitesDisabledUntil,
+                dms_disabled_until = dmsDisabledUntil,
+            });
+
+            // set actions
+            var setActionsResponse = await SecurityActionHelpers.SetCurrentSecurityActions(ctx.Guild.Id, newSecurityActions);
+
+            // respond
+            if (setActionsResponse.IsSuccessStatusCode)
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully paused DMs for **{TimeHelpers.TimeToPrettyFormat(t.Subtract(ctx.Interaction.CreationTimestamp.DateTime.ToLocalTime()), false)}**!");
```
Comment (Erisa) [inline Commands/InteractionCommands/SecurityActionInteractions.cs:45]: Why not Discord timestamps? PrettyFormat has a bunch of issues with time skew and I'm trying not to use it too much.
```suggestion
                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully paused DMs until <t:{TimeHelpers.ToUnixTimestamp(t)}>!");
```

File: Commands/SecurityActions.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+            // set actions
+            var setActionsResponse = await SecurityActionHelpers.SetCurrentSecurityActions(ctx.Guild.Id, newSecurityActions);
+
+            // respond
+            if (setActionsResponse.IsSuccessStatusCode)
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully paused DMs for **{TimeHelpers.TimeToPrettyFormat(t.Subtract(ctx.Message.Timestamp.DateTime), false)}**!");
+            else
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to pause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
+        }
```
Comment (Erisa) [inline Commands/SecurityActions.cs:64]: Lets log this?
```suggestion
            else
            {
                ctx.Client.Logger.LogError("Failed to set Security Actions.\nPayload: {payload}\nResponse: {statuscode} {body}", newSecurityActions.ToString(), (int)setActionsResponse.StatusCode, await setActionsResponse.Content.ReadAsStringAsync());
                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to pause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
            }
```

File: Commands/SecurityActions.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            
+            // set actions
+            var setActionsResponse = await SecurityActionHelpers.SetCurrentSecurityActions(ctx.Guild.Id, newSecurityActions);
+
+            // respond
+            if (setActionsResponse.IsSuccessStatusCode)
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Success} Successfully unpaused DMs!");
+            else
+                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to unpause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
+        }
```
Comment (Erisa) [inline Commands/SecurityActions.cs:109]: Lets log this?
```suggestion
            else
            {
                ctx.Client.Logger.LogError("Failed to set Security Actions.\nPayload: {payload}\nResponse: {statuscode} {body}", newSecurityActions.ToString(), (int)setActionsResponse.StatusCode, await setActionsResponse.Content.ReadAsStringAsync());
                await ctx.RespondAsync($"{Program.cfgjson.Emoji.Error} Something went wrong and I wasn't able to unpause DMs! Discord returned status code `{setActionsResponse.StatusCode}`.");
            }
```

File: Helpers/SecurityActionHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+namespace Cliptok.Helpers
+{
+    public class SecurityActionHelpers
+    {
+        public static async Task<JToken> GetCurrentSecurityActions(ulong guildId)
+        {
+            using HttpRequestMessage getActionsRequest = new(HttpMethod.Get, $"https://discord.com/api/v9/guilds/{guildId}");
```
Comment (Erisa) [inline Helpers/SecurityActionHelpers.cs:7]: Lets make this dynamic so when we do upgrade API version they're all done at the same time.
```suggestion
            using HttpRequestMessage getActionsRequest = new(HttpMethod.Get, $"https://discord.com/api/v{Program.discord.GatewayVersion}/guilds/{guildId}");
```

File: Helpers/SecurityActionHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            
+            var getActionsResponse = await Program.httpClient.SendAsync(getActionsRequest);
+            return ((JObject)JsonConvert.DeserializeObject(await getActionsResponse.Content.ReadAsStringAsync()))["incidents_data"];
+        }
+
+        public static async Task<HttpResponseMessage> SetCurrentSecurityActions(ulong guildId, string newSecurityActions)
+        {
+            // create & send request
+            
+            using HttpRequestMessage setActionsRequest = new(HttpMethod.Put, $"https://discord.com/api/v9/guilds/{guildId}/incident-actions");
```
Comment (Erisa) [inline Helpers/SecurityActionHelpers.cs:18]: Lets make this dynamic so when we do upgrade API version they're all done at the same time.
```suggestion
            using HttpRequestMessage setActionsRequest = new(HttpMethod.Put, $"https://discord.com/api/v{Program.discord.GatewayVersion}/guilds/{guildId}/incident-actions");
```

**Review by Erisa — CHANGES_REQUESTED** [review]

File: Commands/InteractionCommands/SecurityActionInteractions.cs (hunk 3 above)
Comment (FloatingMilkshake) [reply Commands/InteractionCommands/SecurityActionInteractions.cs:45]: Good question! The first thing that came to mind for me was how we show times in warnings and stuff, which uses PrettyFormat. Discord timestamps are a better idea!

**Review by Erisa — APPROVED** [review]

**Pull request merged**